### PR TITLE
[SYCL][ESIMD] Rework -O0 behavior

### PR DESCRIPTION
**Pull request by sarnex**
Let me first say I don't like this change and I wish we didn't have to do this, and it should only be temporary.

The root need is that the IGC VC backend does not support O0 code. It gets miscompiled and produces wrong answers and crashes. The team acknowledges this issue but does not have the resources to fix this at the moment.

From the ESIMD user POV, this is really frustrating. We decided it would be okay if they lost debuggability of device code (but retained it for host code) as long as their program runs and they get the right answer.

So the overall approach is to optimize ESIMD(Not SYCL!!) code even in O0 mode. We are actually already doing this, but this change extends it a bit.

So there are a few main changes in this PR:

1) Don't pass `-O0` to `sycl-post-link`. Note this option has absolutely no effect on SYCL code today, it is only checked in the ESIMD lowering code in `sycl-post-link`.
2) Add a new pass to remove `optnone`/`noinline` early for ESIMD code in `O0`.
3) Remove `optnone`/`noinline` from all functions if we split out ESIMD/SYCL code, and if we don't don't touch it.

All three of these steps are required to fix the tests failing with `O0`, if we remove any one of them we see many failures.

When running all ESIMD tests with `O0`, we see the following improvements:
```
Linux:
Today: 97 failures
With this change: 23 failures, with ~3 root causes. There are no regressions. I'll look into these later.
```
```
Windows: 
Today: 38 failures
With this change: 10 failures. There are no regressions. I'll look into these later.
```

Once IGC VC can actually correctly compile non-optimized code, we should revert this change.

### CODE REVIEW

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -955,6 +955,8 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       PB.registerPipelineStartEPCallback([&](ModulePassManager &MPM,
                                              OptimizationLevel Level) {
         MPM.addPass(ESIMDVerifierPass(LangOpts.SYCLESIMDForceStatelessMem));
+        if (Level == OptimizationLevel::O0)
+          MPM.addPass(ESIMDRemoveOptnoneNoinlinePass());
```
Comment (premanandrao) [inline clang/lib/CodeGen/BackendUtil.cpp:959]: I am okay with this ESIMD change.

**Review by premanandrao — APPROVED** [review]
FE changes are okay with me.

**Review by mdtoguchi — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by sarnex** [comment]
@intel/dpcpp-esimd-reviewers @intel/dpcpp-tools-reviewers Ping, thanks!

**Review by maksimsab — APPROVED** [review]
LGTM.

File: llvm/lib/SYCLLowerIR/ESIMD/ESIMDRemoveOptnoneNoinline.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//=------------------------------------------------------------------------=//
+// The GPU backend for ESIMD does not support debugging and we often see crashes
+// or wrong answers if we do not optimize. Remove optnone and noinline from
+// ESIMD functions so users at least can run their programs and get the
+// right answer.
+//=------------------------------------------------------------------------=//
+
+#define DEBUG_TYPE "ESIMDRemoveOptnoneNoinlinePass"
```
[the remark this answers is not in the document]
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/ESIMDRemoveOptnoneNoinline.cpp:14]: Thanks for the review, I'll remove it for now.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 3]
```diff
@@ -1834,6 +1834,15 @@ bool SYCLLowerESIMDPass::prepareForAlwaysInliner(Module &M) {
       continue;
     }
 
+    // If we are splitting by ESIMD, we are guarenteed the entire
```
Comment (v-klochkov) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:1837]: Is this code still needed after adding the special pass that remove those 2 attributes?

Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:1837]: Yes, I tested many combinations of with/without different changes in this PR and everything we currently have is required.

**Review by v-klochkov — APPROVED** [review]

**Comment by sarnex** [comment]
Windows fail is not related

**Pull request merged**